smarts/core/plan.py

```diff
@@ -0,0 +1,332 @@
+    pass
+
+
+# XXX: consider using smarts.core.coordinates.Pose for this
+@dataclass(frozen=True)
+class Start:
+    position: Tuple[int, int]
+    heading: Heading
+    from_front_bumper: Optional[bool] = True
+
+    @property
+    def point(self) -> Point:
+        return Point(*self.position)
+
+    @classmethod
+    def from_pose(cls, pose: Pose):
+        return cls(
+            position=pose.position[:2],
+            heading=pose.heading,
+            from_front_bumper=False,
+        )
+
+
+@dataclass(frozen=True)
+class Goal:
+    def is_endless(self) -> bool:
+        return True
+
+    def is_reached(self, vehicle) -> bool:
+        return False
+
+
+@dataclass(frozen=True)
+class EndlessGoal(Goal):
+    pass
+
+
+@dataclass(frozen=True)
+class PositionalGoal(Goal):
+    position: Point
+    # target_heading: Heading
+    radius: float
+
+    @classmethod
+    def from_road(
+        cls,
+        road_id: str,
+        road_map: RoadMap,
+        lane_index: int = 0,
+        lane_offset: float = None,
+        radius: float = 1,
+    ):
+        road = road_map.road_by_id(road_id)
+        lane = road.lane_at_index(lane_index)  # XXX: bidirectional roads?
+
+        if lane_offset is None:
+            # Default to the midpoint safely ensuring we are on the lane and not
+            # bordering another
+            lane_offset = lane.length * 0.5
+
+        position = lane.from_lane_coord(RefLinePoint(lane_offset))
+        return cls(position=position, radius=radius)
+
+    def is_endless(self) -> bool:
+        return False
+
+    def is_reached(self, vehicle) -> bool:
+        a = vehicle.position
+        b = self.position
+        dist = math.sqrt((a[0] - b.x) ** 2 + (a[1] - b.y) ** 2)
+        return dist <= self.radius
+
+
+@dataclass(frozen=True)
+class TraverseGoal(Goal):
+    """A TraverseGoal is satisfied whenever an Agent-driven vehicle
+    successfully finishes traversing a non-closed (acyclical) map
+    It's a way for the vehicle to exit the simulation successfully,
+    for example, driving across from one side to the other on a
+    straight road and then continuing off the map.  This goal is
+    non-specific about *where* the map is exited, save for that
+    the vehicle must be going the correct direction in its lane
+    just prior to doing so."""
+
+    def __init__(self, road_map: RoadMap):
+        super().__init__()
+        self._road_map = road_map
+
+    def is_endless(self) -> bool:
+        return True
+
+    def is_reached(self, vehicle) -> bool:
+        return self._drove_off_map(vehicle.position, vehicle.heading)
+
+    def _drove_off_map(
+        self, veh_position: Tuple[float, float, float], veh_heading: float
+    ) -> bool:
+        # try to determine if the vehicle "exited" the map by driving beyond the end of a dead-end lane.
+        nearest_lanes = self._road_map.nearest_lanes(veh_position)
+        if not nearest_lanes:
+            return False  # we can't tell anything here
```

Comment:
Yeah, this method worked as intended before this refactor.  However, since it's low-priority (only used by imitation learning afaik), I left it until "last" to convert.  It's one of the loose ends I plan to clean up on my final pass before any merge.